packages/react/src/primitives/shared/responsive/useBreakpoint.ts

```diff
@@ -0,0 +1,57 @@
+// Inspiration for getMediaQueries and useBreakpoint
+// comes from https://github.com/iiroj/use-breakpoint/
+
+import * as React from 'react';
+import { Breakpoint, UseBreakpoint } from '../../types/responsive';
+import { getMediaQueries } from './getMediaQueries';
+
+const useIsomorphicEffect =
+  typeof window === 'undefined' ? React.useEffect : React.useLayoutEffect;
+
+export const useBreakpoint: UseBreakpoint = ({
+  breakpoints,
+  breakpointUnit,
+  defaultBreakpoint,
+}) => {
+  const supportMatchMedia =
+    typeof window !== 'undefined' && typeof window.matchMedia !== 'undefined';
+  const matchMedia = supportMatchMedia ? window.matchMedia : null;
+
+  const mediaQueries = React.useMemo(
+    () => getMediaQueries({ breakpoints, breakpointUnit }),
+    [breakpoints]
+  );
+
+  const [breakpoint, setBreakpoint] =
+    React.useState<Breakpoint>(defaultBreakpoint);
```

Comment:
@ericclemmons Do you see any issues with server side rendering a default breakpoint from a NextJS perspective? I wasn't able to find any issues testing with the next example app, but I don't have a lot of NextJS experience.

Reply:
Yes will be a problem.

The general rule of thumb for isomorphic apps is **prevent mismatches between what the server & client renders**.

Because the server is mostly unaware of the client dimensions (until responsive headers are adopted, or if the server guesses based on User Agent), this approach often has a responsive-equivalent of [FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content#:~:text=A%20flash%20of%20unstyled%20content,before%20all%20information%20is%20retrieved.):

1. Server defaults to some style (in this case `flex-direction: column`) & sends response

    ```html
    <div class="amplify-flex" style="align-items:stretch;flex-direction:column;padding:0 0 3.0rem 0">
	```

2. Browser renders content + styles

	> <img width="1479" alt="Screen Shot 2021-09-10 at 9 13 24 AM" src="https://user-images.githubusercontent.com/15182/132869286-c9281484-ffe2-4743-8e7a-e4f710108438.png">


3. More styles are loaded or computed (in this case `flex-direction: row`)

	```html
	<div class="amplify-flex" style="align-items: stretch; flex-direction: row; padding: 0px 0px 3rem;">
	```

4. Major layout shifts & redraws happen (as evident on non-highend devices):

	> <img width="1479" alt="Screen Shot 2021-09-10 at 9 13 36 AM" src="https://user-images.githubusercontent.com/15182/132869334-3aa68f15-64d5-4e40-9b8d-4b4ed7d2b9ba.png">


Here's a full vid:

https://user-images.githubusercontent.com/15182/132869164-e8cc57cb-631e-4cc2-9765-72aa520a3e2f.mp4


**So what's the best way of doing this?** IME, it's to ensure the server sends **everything** the client needs to correctly render a responsive the first time.

[Tailwind's use of responsive breakpoints](https://tailwindcss.com/docs/responsive-design) (e.g. `sm:`, `lg:`) work great because all the styles exist up-front.

[Vanilla Extract uses static extraction](https://vanilla-extract.style/documentation/setup/#nextjs) for responsive styles: https://vanilla-extract.style/documentation/styling-api/#assignvars

I would recommend for us that we **conditionally apply classes (e.g. `lg:flex-row`) with matching CSS rather than dynamically setting `style`**.

Reply:
Thanks for the info! Responsive support is an area that will definitely need to be iterated upon. I'll add a ticket to investigate this further.